Support for `nerdctl build` on Windows

TBBle opened the pull request:
Trivially bring up BuildKit support in nerdctl. It was basically all-there already, just disabled.

Resolves #627 assuming you have Buildkitd functional on Windows, of course.

Resolves #2173 which just turned out to be a warning that no error had occurred.

Tested with a trivial Dockerfile that was already working with `buildctl build` and `nerdctl run`, and buildkitd set up as a service matching [the existing rootful instructions](https://github.com/containerd/nerdctl/blob/main/docs/build.md#rootful) (i.e. configured to be in the `default` containerd namespace instead of the default `buildkit` containerd namespace) per https://github.com/moby/buildkit/issues/616#issuecomment-1773975605.

```
nerdctl build --buildkit-host "npipe:////./pipe/buildkitd" . -t docker.io/tbble/supersimpledocker
nerdctl build . -t docker.io/tbble/supersimpledocker
```

Possible outstanding things:

Removing the CLI defaults, while correct because they were not actually used, isn't super-nice from the UX perspective. It might be better to report which paths will be auto-detected, but that requires some further refactoring, and might also be too noisy? Simply saying `<AUTODETECTED>` or something would work, perhaps.

I've only tested this on Windows, and haven't checked if the test suite covers the buildkit socket detection code on Linux/FreeBSD, so it's possible I mangled something when moving that code. And I only tested the default namespace, but since Windows buildkitd doesn't change its default listening address for different containerd namespaces, this seems fine for now.

Actually, the buildkitd source doesn't appear to use different sockets for different containerd namespaces on Linux either, AFAICT. Is that logic actually syncing with a systemd unit configuration or something, rather than buildkitd's logic? It would make sense for buildkitd to work this way, at least for the containerd worker, it just doesn't seem to, or I didn't find the implementation when I went looking.

TBBle commented:
Okay, I broke something when moving the socket-finding logic, the rootless tests are panicking. I'll look into that later today.

slonopotamus commented on pkg/buildkitutil/buildkitutil_unix.go:
```diff
@@ -0,0 +1,50 @@
+		return []string{}, fmt.Errorf("namespace must be specified")
+	}
+	// Try candidate locations of the current containerd namespace.
+	run := "/run/"
```
Ain't it `/var/run` on FreeBSD?

TBBle replied:
I have no idea, I just moved this code block over from the other file. If you'd asked me yesterday, I'd have said it was `/var/run` on Linux too, having grown up in a pre-FHS3.0 world.

AkihiroSuda replied:
`/var/run` seems safer on non-Linux

slonopotamus commented on pkg/buildkitutil/buildkitutil_unix.go:
```diff
@@ -0,0 +1,50 @@
+//go:build freebsd || linux
```
`//go:build unix` maybe?

TBBle replied:
I just copied this from another pair of files that was split between `_windows.go` and `_unix.go`. Looking around, I see this was added in Go 1.19, but not added for filenames, which is mildly annoying, but I understand the rationale.

So yeah, that makes sense, as long as we're not expecting to support go 1.18. It's not yet used in the codebase, but I did see one `!windows` which I had never thought of... So probably worth fixing project-wide.

TinaMor commented on pkg/buildkitutil/buildkitutil.go:
```diff
@@ -144,8 +130,8 @@ func PingBKDaemon(buildkitHost string) error {
 func pingBKDaemon(buildkitHost string) (output string, _ error) {
-	if runtime.GOOS != "linux" && runtime.GOOS != "freebsd" {
-		return "", errors.New("only linux and freebsd are supported")
+	if runtime.GOOS != "linux" && runtime.GOOS != "freebsd" && runtime.GOOS != "windows" {
```
Nit: Easier to maintain
```
import "slices"
...
supportedOs := []string{"linux", "freebsd", "windows"}
if !slices.Contains(supportedOs , runtime.GOOS)) {
...
}
```

TBBle replied:
Sadly, this turns out to require Go 1.21 or newer for the "slices" package. I'll open-code slices.Contains for now, and leave a note that we cam replace it with "slices" later.

AkihiroSuda commented:
Needs rebase

iankingori commented:
@TBBle do you have some free cycles to complete this? [buildkit v0.13.0-rc2](https://github.com/moby/buildkit/releases/tag/v0.13.0-rc2) includes the windows buildkitd binary and would be great to have nerdctl working as well. Happy to pick this up if you are swamped

TBBle commented:
I can quickly rebase it tonight, as it looks like the conflicting changes are simply adjacent to my changes, rather than actual conflicting changes. The requested changes don't look too bad, so I can probably get those done too tonight.

TBBle commented:
Okay, I've rebased it, and we'll let CI confirm I didn't break anything. I am not currently in a position to test it locally tonight though.

Also, I added https://github.com/containerd/nerdctl/pull/2587/commits/58625bf1559a613380ba4291b8d86ad38207eb4c for https://github.com/containerd/nerdctl/pull/2587#pullrequestreview-1691472221 but it's mildly out-of-scope and slightly touches a handful of files so happy to have that split out into a separate PR. That one also makes Go 1.19 the minimum, but go.mod already states Go 1.20 as minimum so I assume that's probably fine.

iankingori commented:
> Okay, I've rebased it, and we'll let CI confirm I didn't break anything. I am not currently in a position to test it locally tonight though.

I should have some time to test this today, thanks.

iankingori commented:
I got `nerdctl build` to run but with a couple of problems:

**1.Nerdctl sets a default windows path on container images:** 

```set PATH
Path=c:\Windows\System32;c:\Windows
```

 This is a universal problem we've seen with Buildkit as well that we need to address but unrelated to this PR.

**2.Nerdctl Build does not work when progress is `auto` or `tty`**

When `progress=auto` which defaults to tty we get the following:

**Nerdctl**

 Command: ```nerdctl build . -t core:1```
```
time="2024-02-28T23:10:18-08:00" level=debug msg="Choosing the buildkit host \"npipe:////./pipe/buildkitd\", candidates=[npipe:////./pipe/buildkitd]"
time="2024-02-28T23:10:18-08:00" level=debug msg="Chosen buildkit host \"npipe:////./pipe/buildkitd\""
time="2024-02-28T23:10:19-08:00" level=debug msg="worker labels: map[org.mobyproject.buildkit.worker.containerd.namespace:buildkit org.mobyproject.buildkit.worker.containerd.uuid:6ceace98-92bc-4a83-807b-133ad2fd9275 org.mobyproject.buildkit.worker.executor:containerd org.mobyproject.buildkit.worker.hostname:WIN-LIQNSMAKV3L org.mobyproject.buildkit.worker.network: org.mobyproject.buildkit.worker.selinux.enabled:false org.mobyproject.buildkit.worker.snapshotter:windows]"
time="2024-02-28T23:10:19-08:00" level=debug msg="running C:\\Tools\\Buildkit\\bin\\buildctl.exe [--addr=npipe:////./pipe/buildkitd build --progress=tty --frontend=dockerfile.v0 --local=context=. --output=type=docker,name=docker.io/library/core:1 --local=dockerfile=C:\\code\\debug --opt=filename=Dockerfile]"
time="2024-02-28T23:10:19-08:00" level=fatal msg="no image was built"
```

**Buildkit**

Response:
```time="2024-02-28T23:13:19-08:00" level=error msg="/moby.buildkit.v1.Control/Solve returned error: rpc error: code = Unknown desc = failed to copy to tar: rpc error: code = Unknown desc = write /dev/stdout: The pipe is being closed."
failed to copy to tar: rpc error: code = Unknown desc = write /dev/stdout: The pipe is being closed.
6532 v0.13.0-rc2 C:\Tools\Buildkit\bin\buildctl.exe --addr=npipe:////./pipe/buildkitd build --progress=tty --frontend=dockerfile.v0 --local=context=. --output=type=docker,name=docker.io/library/core:1 --local=dockerfile=C:\code\debug --opt=filename=Dockerfile
github.com/moby/buildkit/session/filesync.writeTargetFile
        /src/session/filesync/diffcopy.go:145
github.com/moby/buildkit/session/filesync.(*fsSyncAttachable).DiffCopy
        /src/session/filesync/filesync.go:339
github.com/moby/buildkit/session/filesync._FileSend_DiffCopy_Handler
        /src/session/filesync/filesync.pb.go:392
github.com/moby/buildkit/util/grpcerrors.StreamServerInterceptor
        /src/util/grpcerrors/intercept.go:33
github.com/moby/buildkit/session.NewSession.ChainStreamServer.func6.1.1
        /src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:49
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.StreamServerInterceptor.func1
        /src/vendor/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:446
github.com/moby/buildkit/session.NewSession.ChainStreamServer.func6.1.1
        /src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:49
github.com/moby/buildkit/session.NewSession.ChainStreamServer.func6
        /src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:58
google.golang.org/grpc.(*Server).processStreamingRPC
        /src/vendor/google.golang.org/grpc/server.go:1644
google.golang.org/grpc.(*Server).handleStream
        /src/vendor/google.golang.org/grpc/server.go:1741
google.golang.org/grpc.(*Server).serveStreams.func1.1
        /src/vendor/google.golang.org/grpc/server.go:986
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1650
```

This particular issue goes away when we switch to `progress=plain`

**Nerdctl**

Command: ``` nerdctl build --progress=plain . -t core:1```

Response:
```
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 176B done
#1 DONE 0.0s

#2 [internal] load metadata for mcr.microsoft.com/windows/servercore:ltsc2022
#2 DONE 0.1s

#3 [internal] load .dockerignore
#3 transferring context: 2B done
#3 DONE 0.0s

#4 [1/2] FROM mcr.microsoft.com/windows/servercore:ltsc2022@sha256:8a75266be74ad7a904470e18057f6fd62055cf7028172307fefc67aff37dfd10
#4 resolve mcr.microsoft.com/windows/servercore:ltsc2022@sha256:8a75266be74ad7a904470e18057f6fd62055cf7028172307fefc67aff37dfd10 0.0s done
#4 DONE 0.1s

#4 [1/2] FROM mcr.microsoft.com/windows/servercore:ltsc2022@sha256:8a75266be74ad7a904470e18057f6fd62055cf7028172307fefc67aff37dfd10
#4 extracting sha256:d7662b0a97ab33c03d2f9da1cbaf9628276b35d8445cec2791dd15133452c1d1
#4 extracting sha256:d7662b0a97ab33c03d2f9da1cbaf9628276b35d8445cec2791dd15133452c1d1 82.4s done
#4 DONE 82.5s

#4 [1/2] FROM mcr.microsoft.com/windows/servercore:ltsc2022@sha256:8a75266be74ad7a904470e18057f6fd62055cf7028172307fefc67aff37dfd10
#4 extracting sha256:4abd29fcbfc570cbbbaefefc279496e8f41ab12aa0e2e608ffd71916ba664a81
#4 extracting sha256:4abd29fcbfc570cbbbaefefc279496e8f41ab12aa0e2e608ffd71916ba664a81 91.0s done
#4 DONE 173.5s

#5 [2/2] RUN "echo Hello Nerdctl!!"
#5 1.755 Hello Nerdctl!!
#5 DONE 2.7s

#6 exporting to docker image format
#6 exporting layers
#6 exporting layers 2.4s done
#6 exporting manifest sha256:318cf6a359222644681c0d0b561a939a70fa6bcec90fe152724b1235bb129df9 0.0s done
#6 exporting config sha256:57db8b05d1bb3c16224cc694ac90a87c05f1f27a7ef19886baa0056a9009cbfb 0.0s done
#6 sending tarball
#6 sending tarball 36.9s done
#6 DONE 39.3s
unpacking docker.io/library/core:1 (sha256:318cf6a359222644681c0d0b561a939a70fa6bcec90fe152724b1235bb129df9)...
Loaded image: docker.io/library/core:1
```

**Short Term Fix:** 
We could default windows progress to plain and/or log this as a know issue

**Long Term Fix:** 
Investigate how to pass the Windows Stdout Handle from Nerdctl to Buildkit. I suspect we need to separate the progress text stream from the tar file stream.

TBBle commented:
Huh, weird, I didn't see the TTY issue in my original testing. I wonder if something changed since then, or if my test setup was so trivial it didn't trigger this. If I recall correctly, my test setup here didn't actually have anything in the context except the Dockerfile.

AkihiroSuda commented:
Can we merge this and track other issues in follow-up ?

iankingori commented:
Sounds good to me

TBBle commented on pkg/buildkitutil/buildkitutil.go:
```diff
@@ -143,9 +130,21 @@ func PingBKDaemon(buildkitHost string) error {
 	return nil
 }
 
+// contains open-codes slices.Contains (without generics) from Go 1.21.
```
Nit: Typo: "open-coded". >_<

AkihiroSuda replied:
Do you plan to fix this before merging?

TBBle replied:
Ooops, looking at it again, the comment is actually correct: `contains` is the name of the function.

AkihiroSuda reviewed (approved):
Thanks

slonopotamus commented:
Doesn't seem to work, see https://github.com/containerd/containerd/issues/10154#issuecomment-2222472087

TBBle commented:
Oops, looks like we never opened an issue for number 2 in https://github.com/containerd/nerdctl/pull/2587#issuecomment-1970588592.
